### PR TITLE
Align support for negative session ids

### PR DESCRIPTION
**Pull request by sjanusz-r7**
### Before

The syntax `sessions -u -1` isn't supported:
```
msf6 exploit(windows/smb/ms17_010_eternalblue) > sessions -u -1
[-] Please specify valid session identifier(s)
```

The syntax `jobs -k -1` isn't supported:
```
msf6 auxiliary(server/socks_proxy) > jobs -k -1
[-] Please specify valid job identifier(s)
```

### After

The last session is now upgraded as expected:
```
msf6 auxiliary(scanner/ssh/ssh_login) > sessions -u -1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [-1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.28.128.1:6969 
[-] Powershell is not installed on the target.
...etc...
```

The last job is now killed as expected:
```
msf6 auxiliary(server/socks_proxy) > jobs -k -1
[*] Stopping the following job(s): -1
[*] Stopping job 1
```

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use auxiliary/server/socks_proxy`
- [ ] `run -j`
- [ ] `jobs -k -1`
- [ ] The last job will be killed.
- [ ] `use auxiliary/scanner/ssh/ssh_login`
- [ ] `run`
- [ ] Background the session and run `sessions -u -1`
- [ ] The last session will be upgraded.

### CODE REVIEW

File: lib/msf/ui/console/command_dispatcher.rb [hunk 1]
```diff
@@ -102,10 +103,16 @@ def build_range_array(id_list)
...
       temp_list = id_list.split(',')
       temp_list.each do |ele|
-        return if ele.count('-') > 1
-        return if ele.first == '-' || ele[-1] == '-'
-        return if ele.first == '.' || ele[-1] == '.'
-        return unless ele =~ (/^\d+((\.\.|-)\d+)?$/)  # Not a number or range
+        return if ele.count('-') > 1 # Eg. 'sessions -u -1-,5', incorrect syntax
+        return if ele[-1] == '-' # Last item of array is a '-', resulting in an incomplete range
+        return if ele.first == '.' || ele[-1] == '.' #Eg. 'sessions -u .1..' or 'sessions -u ..
+        return unless ele =~ (/^\d+((\.\.|-)\d+)?$/) || ele =~ (/^((-?\d+),?)+$/) # Not a number or range
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher.rb:109]: Looks like this regex doesn't need to handle `,` as it's already split up

File: lib/msf/ui/console/command_dispatcher.rb [hunk 2]
```diff
@@ -102,10 +103,16 @@ def build_range_array(id_list)
...
+
+        # Check if the item is negative, as this will not always be a range
+        if (ele.to_i < 0)
+          item_list.push(ele.to_i)
+          next
+        end
 
         if ele.include? '-'
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher.rb:117]:
```suggestion
        elsif ele.include? '-'
```

File: lib/msf/ui/console/command_dispatcher.rb [hunk 3]
```diff
@@ -102,10 +103,16 @@ def build_range_array(id_list)
...
-        return if ele.first == '-' || ele[-1] == '-'
-        return if ele.first == '.' || ele[-1] == '.'
-        return unless ele =~ (/^\d+((\.\.|-)\d+)?$/)  # Not a number or range
+        return if ele.count('-') > 1 # Eg. 'sessions -u -1-,5', incorrect syntax
+        return if ele[-1] == '-' # Last item of array is a '-', resulting in an incomplete range
+        return if ele.first == '.' || ele[-1] == '.' #Eg. 'sessions -u .1..' or 'sessions -u ..
+        return unless ele =~ (/^\d+((\.\.|-)\d+)?$/) || ele =~ (/^((-?\d+),?)+$/) # Not a number or range
+
+        # Check if the item is negative, as this will not always be a range
+        if (ele.to_i < 0)
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher.rb:112]:
```suggestion
        if ele.to_i < 0
```

File: lib/msf/ui/console/command_dispatcher/jobs.rb [hunk 4]
```diff
@@ -146,8 +146,12 @@ def cmd_jobs(*args)
...
               when "-p"
                 job_list = build_range_array(val)
-                job_list.each do |job_id|
-                  add_persist_job(job_id)
+                if !job_list.blank?
+                  job_list.each do |job_id|
+                    add_persist_job(job_id)
+                  end
+                else
+                  add_persist_job(nil)
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher/jobs.rb:154]:
```suggestion
                job_list.each do |job_id|
                  add_persist_job(job_id)
                end
```

Just confirming; I thought we were updating the implementation to always return an array? 👀

File: lib/msf/ui/console/command_dispatcher.rb [hunk 5]
```diff
@@ -93,7 +93,8 @@ def docs_dir
 
   #
   # Generate an array of job or session IDs from a given range String.
-  # Always returns an Array.
+  # Always returns an Array unless an incorrect input is given.
+  # In that case, the result will always be nil, even if only one argument is incorrect.
   #
   # @param id_list [String] Range or list description such as 1-5 or 1,3,5 etc
   # @return [Array<String>] Representing the range
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher.rb:100]:
```suggestion
  # @return [Array<String>, nil] Representing the range
```

File: spec/lib/msf/ui/console/command_dispatcher_spec.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+      {id_list: '2,-1', expected: [-1, 2]},
+      {id_list: '-1,-2', expected: [-2, -1]},
+      {id_list: '-1-', expected: nil},
+      {id_list: '-1-,2', expected: nil},
+      {id_list: '-1--,2', expected: nil},
+      {id_list: '---1', expected: nil},
+      {id_list: '1--', expected: nil},
+      {id_list: '1-3', expected: [1, 2, 3]}
+    ].each do |test|
+      it "returns #{test[:expected]} for the id_list #{test[:id_list]}" do
```
Comment (adfoster-r7) [inline spec/lib/msf/ui/console/command_dispatcher_spec.rb:30]:
```suggestion
      it "returns #{test[:expected].inspect} for the id_list #{test[:id_list]}" do
```

Comment (adfoster-r7) [reply spec/lib/msf/ui/console/command_dispatcher_spec.rb:30]: That way `nil` appears in the description:
```ruby
>> "expected: #{nil}"
=> "expected: "
>> "expected: #{nil.inspect}"
=> "expected: nil"
```

File: spec/lib/msf/ui/console/command_dispatcher_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+  let(:subject) do
+    dummy_class = Class.new
+    dummy_class.include described_class
+    dummy_class.new(driver)
+  end
+
+  describe '#build_range_array' do
+    [
+      {id_list: '1', expected: [1]},
```
Comment (adfoster-r7) [inline spec/lib/msf/ui/console/command_dispatcher_spec.rb:15]: 💎 👮‍♂️ 🙏

File: spec/lib/msf/ui/console/command_dispatcher_spec.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      {id_list: '1,2', expected: [1, 2]},
+      {id_list: '-1,2', expected: [-1, 2]},
+      {id_list: '2,-1', expected: [-1, 2]},
+      {id_list: '-1,-2', expected: [-2, -1]},
+      {id_list: '-1-', expected: nil},
+      {id_list: '-1-,2', expected: nil},
+      {id_list: '-1--,2', expected: nil},
+      {id_list: '---1', expected: nil},
+      {id_list: '1--', expected: nil},
+      {id_list: '1-3', expected: [1, 2, 3]}
```
Comment (adfoster-r7) [inline spec/lib/msf/ui/console/command_dispatcher_spec.rb:28]:
```suggestion
      { input : '1-3', expected: [1, 2, 3] }
```

File: lib/msf/ui/console/command_dispatcher.rb [hunk 9]
```diff
@@ -93,21 +93,25 @@ def docs_dir
...
     item_list = []
     unless id_list.blank?
       temp_list = id_list.split(',')
       temp_list.each do |ele|
-        return if ele.count('-') > 1
-        return if ele.first == '-' || ele[-1] == '-'
-        return if ele.first == '.' || ele[-1] == '.'
-        return unless ele =~ (/^\d+((\.\.|-)\d+)?$/)  # Not a number or range
+        return if ele.count('-') > 1 # Eg. 'sessions -u -1-,5', incorrect syntax
+        return if ele[-1] == '-' # Last item of array is a '-', resulting in an incomplete range
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher.rb:107]:
```suggestion
        return if ele.last == '-' # Last item of array is a '-', resulting in an incomplete range
```

File: lib/msf/ui/console/command_dispatcher.rb [hunk 10]
```diff
@@ -93,21 +93,25 @@ def docs_dir
...
-        return if ele.first == '.' || ele[-1] == '.'
-        return unless ele =~ (/^\d+((\.\.|-)\d+)?$/)  # Not a number or range
+        return if ele.count('-') > 1 # Eg. 'sessions -u -1-,5', incorrect syntax
+        return if ele.last == '-' # Last item of array is a '-', resulting in an incomplete range
+        return if ele.first == '.' || ele.last == '.' #Eg. 'sessions -u .1..' or 'sessions -u ..
+        return unless ele =~ (/^-?\d+((\.\.|-)\d+)?$/) # Not a number or range
 
-        if ele.include? '-'
+        # Check if the item is negative, as this will not always be a range
+        if ele.to_i < 0
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher.rb:112]: I think this breaks for:
```
      { input: '-1..3', expected: [-1, 0, 1, 2, 3] }
```

```
       expected: [-1, 0, 1, 2, 3]
            got: [-1]

**Comment by smcintyre-r7** [comment]
I like the `-1` support to reference jobs! That's a great idea/addition.

**Comment by adfoster-r7** [comment]
@msjenkins-r7 retest this please

File: lib/msf/ui/console/command_dispatcher/jobs.rb [hunk 11]
```diff
@@ -146,6 +146,10 @@ def cmd_jobs(*args)
                 job_id = val
               when "-p"
                 job_list = build_range_array(val)
+                if job_list.blank?
+                  print_error('Invalid job ID')
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher/jobs.rb:150]:
```suggestion
                  print_error('Please specify valid job identifier(s)')
```

**Comment by adfoster-r7** [comment]
Working as expected:
```
msf6 exploit(unix/webapp/thinkphp_rce) > sessions -u -1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [-1]

[*] Upgrading session ID: 5
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 192.168.123.1:4433 

[*] Sending stage (984904 bytes) to 192.168.123.6
[*] Meterpreter session 6 opened (192.168.123.1:4433 -> 192.168.123.6:54022) at 2021-07-30 10:16:45 +0100
[*] Command stager progress: 100.00% (773/773 bytes)
```

```
msf6 auxiliary(server/socks_proxy) > jobs -k -1
[*] Stopping the following job(s): -1
[*] Stopping job 1

[*] Stopping the SOCKS proxy server
```

**Comment by jmartin-r7** [comment]
## Release notes
This expands usability for job interactions by enabling job id `-1` to reference the most recently created job.

**Pull request merged**